conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/WebSecurityHandler.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.conjure.java.undertow.runtime;
+
+import com.palantir.logsafe.Preconditions;
+import io.undertow.server.ConduitWrapper;
+import io.undertow.server.HttpHandler;
+import io.undertow.server.HttpServerExchange;
+import io.undertow.util.ConduitFactory;
+import io.undertow.util.HeaderMap;
+import io.undertow.util.Headers;
+import io.undertow.util.HttpString;
+import org.xnio.conduits.StreamSinkConduit;
+
+final class WebSecurityHandler implements HttpHandler {
+
+    private static final String CONTENT_SECURITY_POLICY =
+            "default-src 'self'; img-src 'self' data:; style-src 'self' 'unsafe-inline'; frame-ancestors 'self';";
+    private static final String CONTENT_TYPE_OPTIONS = "nosniff";
+    private static final String FRAME_OPTIONS = "sameorigin";
+    private static final String XSS_PROTECTION = "1; mode=block";
+
+    private static final String REFERRER_POLICY = "strict-origin-when-cross-origin";
+
+    private static final HttpString HEADER_IE_X_CONTENT_SECURITY_POLICY =
+            HttpString.tryFromString("X-Content-Security-Policy");
+    private static final String USER_AGENT_IE_10 = "MSIE 10";
+    private static final String USER_AGENT_IE_11 = "rv:11.0";
+
+    private final HttpHandler next;
+
+    WebSecurityHandler(HttpHandler next) {
+        this.next = Preconditions.checkNotNull(next, "Delegate HttpHandler is required");
+    }
+
+    @Override
+    public void handleRequest(HttpServerExchange exchange) throws Exception {
+        exchange.addResponseWrapper(Listener.INSTANCE);
+        next.handleRequest(exchange);
+    }
+
+    private enum Listener implements ConduitWrapper<StreamSinkConduit> {
+        INSTANCE;
+
+        @Override
+        public StreamSinkConduit wrap(ConduitFactory<StreamSinkConduit> factory, HttpServerExchange exchange) {
+            HeaderMap headers = exchange.getResponseHeaders();
+            headers.put(Headers.CONTENT_SECURITY_POLICY, CONTENT_SECURITY_POLICY);
+            headers.put(Headers.REFERRER_POLICY, REFERRER_POLICY);
+
+            String userAgent = exchange.getRequestHeaders().getFirst(Headers.USER_AGENT);
+            if (userAgent != null) {
```

Comment:
[This](https://content-security-policy.com/) page suggests that we may run into trouble setting both `content-security-policy` and `x-content-security-policy`: 

> Note: It is known that having both Content-Security-Policy and X-Content-Security-Policy or X-Webkit-CSP causes unexpected behaviours on certain versions of browsers. Please avoid using deprecated X-* headers.

(But if we haven't seen any issues in practice it may not be worth changing here)

Reply:
was this resolved?

Reply:
based on `if we haven't seen any issues in practice it may not be worth changing here` I'd say so. I prefer not to make changes here unless things are currently broken, since this is just applying existing logic in a new format.